### ISSUE TITLE
Objects instanced in c++ can't be bound to typed gdscript variables

### ISSUE THREAD

**Issue by norton-corbett** [godotengine/godot-cpp#430]
GDNative:
```c++
class MyCustomType : public godot::Node
{
    GODOT_CLASS(MyCustomType, godot::Node);
public:
    void calling();
    void functions();
    void works();
    void fine();
};

// ...

class Manager: public godot::Node
{
    GODOT_CLASS(Manager, godot::Node);
public:
    MyCustomType* make_custom_object()
    {
        auto object = MyCustomType::_new();
        add_child(object);
        return object;
    }
};
```

The .gdns file:
```
[gd_resource type="NativeScript" load_steps=2 format=2]

[ext_resource path="res://bin/test.gdnlib" type="GDNativeLibrary" id=1]

[resource]
class_name = "MyCustomType"
library = ExtResource( 1 )
script_class_name = "MyCustomType"
```

GDScript:
```gdscript
var o = Manager.make_custom_object()

o.calling()
o.functions()
o.works()
o.fine()

var test:MyCustomType = o #error
```
![image](https://user-images.githubusercontent.com/68328892/88051744-03841a00-cb51-11ea-9a22-200c2f70ec38.png)

Also, if the variable is a function parameter, e.g.:
```gdscript
func broken(param:MyCustomType):
    pass

# ...

var o = Manager.make_custom_object()

broken(o)
```
...then no runtime error is produced, but param will be `null`

**Comment by Zylann** [comment]
I'm not sure if that's a problem with C++ bindings. You gave the global class name through the `.gdns`, it doesn't actually matter if it's implemented with C++ bindings or C API. Could be a Godot bug.

**Comment by norton-corbett** [comment]
Good point. I understand GDScript is being re-implemented currently so maybe I'll wait before reporting it.

**Comment by alessandrofama** [comment]
A workaround that works for me is to wrap `Manager.make_custom_object()`  in an autoload script that returns your typed variable. For example:

```
# Test.gd, name "Test" in Autoload

extends Node

func make_custom_object() -> MyCustomType:
	return Manager.make_custom_object()
```

Then it should work I think:

```
var o:MyCustomType

func _ready():
	o = Test.make_custom_object()
```

At least it does for me, but I also have a .gdns for each class, don't know if it makes a difference.

**Comment by Zylann** [comment]
I don't understand how an autoload solves this tbh.
At some point, you are assigning a value `var test: MyCustomType`, which can fail if the cast is invalid, and that will happen with or without an autoload. Unless I misunderstood some strange detail

**Comment by norton-corbett** [comment]
Yeah this autoload workaround doesn't make any difference for me. Are you also using Godot 3.2 stable?

**Comment by alessandrofama** [comment]
Yeah, 3.2.2. after testing it again it seems that it only works for me if the variable is declared outside functions. Sorry, I'm missing some technical knowledge about why it would work or not, still relatively new to GDNative and trying out stuff :)
In my case I have a GDNative class named `EventInstance`, I create an instance here : 

```
godot::EventInstance* Fmod::createInstance(const String path)
{
// stuff 
	godot::EventInstance* eventInstance = godot::EventInstance::_new();
	return eventInstance;
}
```

In GDScript:

```
# Wrapper.gd
extends Node

func create_instance(event_path) -> EventInstance:
	return Fmod.create_instance(event_path)

# Test.gd 
extends Node 
var instance:EventInstance

func _ready():
	instance = Wrapper.create_instance("event:/Music")
```

This works and I can call the methods registered in the ``EventInstance``  class (they also show up in auto complete).

This doesn't work:
```
# Test.gd
func _ready():
	var instance:EventInstance = Wrapper.create_instance("event:/Music")

```

Both `Fmod.gdns` and `Wrapper.gd` are in AutoLoad and created an additional gdns for the EventInstance class (Class Name & Script Class Name are the same).

**Comment by norton-corbett** [comment]
For me the error still occurs regardless of whether or not the assignment is separated from the declaration. I guess it's a bug either way. Since I opened this issue the new GDScript has been released on the master branch but I haven't tested to see if this is still an issue.

**Comment by Zylann** [comment]
It would see that the problem comes down to the fact the NativeScript API in Godot doesn't allow us to provide the names and types appearing in function signatures, which includes return types. That means Godot always sees return values of NativeScript functions as `Variant`, and as such cannot be typed properly.

This is something I wanted for a long time in Godot because using NativeScript in GDScript is very error prone, coders are basically in the dark as no static checks can ever run.

**Comment by LennyPhoenix** [comment]
I think the issue comes from [how the custom objects are instantiated](https://github.com/godotengine/godot-cpp/blob/master/include/core/Godot.hpp#L39).
[The object's parent type is first instantiated](https://github.com/godotengine/godot-cpp/blob/master/include/core/Godot.hpp#L81) (built-in types do not have this problem), [then a new `NativeScript` object is created](https://github.com/godotengine/godot-cpp/blob/master/include/core/Godot.hpp#L57), [set to the correct values](https://github.com/godotengine/godot-cpp/blob/master/include/core/Godot.hpp#L60) and [assigned to the object](https://github.com/godotengine/godot-cpp/blob/master/include/core/Godot.hpp#L84), essentially creating an instance of the custom object.
The problem arises as this `NativeScript` is actually different to the user-created one, meaning they are technically different "types" to GDScript.
The error gives the variable a blank type name, because the type name would be derived from the `NativeScript`'s file name, but because the created `NativeScript` object is not saved, it has no file name and the typename is just left blank.
As for how this can be resolved, it would probably need a better solution to instantiate custom classes from the GDNative API,

**Comment by Zylann** [comment]
Ah true, we had to do this because there is no way the C++ bindings could guess where the `gdns` files are.
To resolve this, the whole GDNative system needs to be rethought to not need those gdns files in the first place IMO. Which I think is going to happen in Godot 4.

**Comment by LennyPhoenix** [comment]
As a temporary workaround (for the workaround), we could add a static property `GODOT_CLASS` macro which can optionally contain a path to a `.gdns` file. In `create_custom_class_instance`, we could check if the class has a `script_path` set, and if so we can load that script instead of creating a new one.

**Comment by Zylann** [comment]
@LennyPhoenix  There are two problems with this:
- It has to be thread-safe. If you do a just-in-time initialization, it has to be guarded by a mutex somehow, and I'm not sure we can store that in a header without requiring an implementation part in the cpp file
- We still need a path to a valid `.gdns` file, and the bindings cannot know where you put them. Either you have to hardcode a path, or ask Godot where it is somehow (again without breaking thread safety)

**Comment by LennyPhoenix** [comment]
I mean that the user could provide the hard-coded path to the `.gdns` file from within the class, but I'm not sure how that would be implemented, we could potentially create a separate function to instantiate a custom class using a path to an existing `.gdns` file for if the user knows where it is located? In my Discord SDK wrapper I make sure that the `.gdns` files are all located in the same folder as the `.gdnlib` file, and I can locate them using the class names, but this breaks as soon as someone tries to rename a file or move it around.

**Comment by Zylann** [comment]
Sounds like it would work, but again that's a big workaround, I believe it should not even be a thing in Godot 4.

**Comment by LennyPhoenix** [comment]
Yeah I agree that `.gdns` files should be removed in Godot 4, but until there is a better solution it might be a good idea to have this workaround so that people can use static typing in GDScript, as there isn't really any other way to do so.

**Comment by LennyPhoenix** [comment]
> Also, I wonder if the fact the library references .gdns files will not produce a cyclic reference. Because .gdns resources reference the library.

The library does not actually reference the `.gdns` file, but because they are stored in the same folder, and the `.gdns` files are named after their corresponding classes, we can dynamically load the `.gdns` files by using the `.gdnlib`'s path, and converting the class name to `snake_case`. Here is [the implementation](https://github.com/LennyPhoenix/GodotDiscordSDK/blob/main/src/util.c#L3).
